[Bugfix-4996] preserve non breaking white spaces

leewesleyv opened the pull request:
Fixes #4996.

The regex is currently not covering all different cases of non-breaking white spaces in the content (in combination with other whitespace characters). I'm not sure if handling this through a regex is the way to go here.

thibaudcolas commented:
👌 have canceled the Travis builds for now, let's have a look at which patterns are / aren't covered tomorrow.

gasman commented:
Looks like https://www.w3.org/TR/html4/struct/text.html#h-9.1 is the definitive list of whitespace characters that should be collapsed - hopefully it's not too hard to express that as a regexp...

leewesleyv commented:
I've updated the regex to:

- Be a little bit more efficient. It was matching on a single space and replacing it with a single space. It now checks for 2 or more consecutive whitespace characters.
- Made the regex more specific, only targetting **tabs**, **line-feed** and **form-feed** characters.
- Added a regex to replace 2 or more consecutive zero-width spaces with 1 zero-width space (collapse them)

gasman commented on wagtail/admin/rich_text/converters/html_to_contentstate.py:
```diff
@@ -12,7 +12,11 @@
+WHITESPACE_RE = re.compile(r'[ \t\n\f]{2,}')
+
+# match more than one consecutive zero-width spaces
+ZERO_WIDTH_SPACE_RE = re.compile(r'[\u200b]{2,}')
```
What's the reason for collapsing zero-width spaces? Is this something that happens in real-world usage (for example, when pasting content from a Word document)?

I can't see any good reason for a document to have multiple zero-width spaces, but I can't see why they would cause harm either - and in this case I'd prefer not to make non-essential changes.

leewesleyv commented on wagtail/admin/rich_text/converters/html_to_contentstate.py:
```diff
@@ -314,8 +318,9 @@ def handle_endtag(self, name):
 
     def handle_data(self, content):
-        # normalise whitespace sequences to a single space
+        # normalise whitespace sequences to a single character
```
If I'm correct, this will also mean that we need to reset the quantifier of the regex to **+** (was **{2,}**).

gasman replied:
I think so, yes.

leewesleyv commented:
@gasman I've resolved your comments! Please let me know what you think.

thibaudcolas commented:
aaand it's merged! Thank you @leewesleyv 🙂. This will be released in v2.5.
